### PR TITLE
Additional guidance around casts:

### PR DESCRIPTION
**Pull request by ZachHenkel**
[Issue #926 for discussion](https://github.com/isocpp/CppCoreGuidelines/issues/926)

- Warn on unnecessary casts
- Avoid casts on arithmetic types
- Add specializations of the rules for Pro-type-reinterpretcast

### CODE REVIEW

File: CppCoreGuidelines.md [hunk 1]
```diff
@@ -11623,14 +11624,14 @@ Casts are widely (mis) used. Modern C++ has constructs that eliminate the need f
...
-
+* Rely on the well defined, safe, implicit conversions between pointer types
 
 ##### Enforcement
 
 * Force the elimination of C-style casts
-* Warn against named casts
 * Warn if there are many functional style casts (there is an obvious problem in quantifying 'many').
 * The [type profile](#Pro-type-reinterpretcast) bans `reinterpret_cast`.
+* Warn against [unnecessary pointer casts](#Pro-type-unnecessarycast).
```
Comment (gdr-at-ms) [inline CppCoreGuidelines.md]: I think we should use a more precise term here.  E.g. "Warn against [identity pointer casts](#Pro-style-identity-cast)."

File: CppCoreGuidelines.md [hunk 2]
```diff
@@ -19397,8 +19399,10 @@ An implementation of this profile shall recognize the following patterns in sour
 
 Type safety profile summary:
 
-* <a name="Pro-type-reinterpretcast"></a>Type.1: Don't use `reinterpret_cast`:
-A strict version of [Avoid casts](#Res-casts) and [prefer named casts](#Res-casts-named).
+* <a name="Pro-type-avoidcasts"></a>Type.1: [Avoid casts](#Res-casts):  
+<a name="Pro-type-reinterpretcast">a. </a>Don't use `reinterpret_cast`; A strict version of [Avoid casts](#Res-casts) and [prefer named casts](#Res-casts-named).  
+<a name="Pro-type-arithmeticcast">b. </a>Don't a `static_cast` for arithmetic types; A strict version of [Avoid casts](#Res-casts) and [prefer named casts](#Res-casts-named).  
+<a name="Pro-type-unnecessarycast">c. </a>Don't use pointer unnecessary pointer casts; A strict version of [Avoid casts](#Res-casts)
```
Comment (gdr-at-ms) [inline CppCoreGuidelines.md]: Suggestion: "<a name="Pro-type-identity-cast">c.<a/> Don't use pointer casts where the source and the target types are the same;"

This suggestion is based on the discussion from last review when I asked "what is unnecessary cast" and I was told that you had in mind "casts where the source and target types are the same."

People have complained in the past that the rules are too imprecise or unimplementable,  I want to ensure that as we add new rule or we refine existing rules, we are making them precise and definitely implementable.

File: CppCoreGuidelines.md (hunk 1 above)
Comment (ZachHenkel) [inline CppCoreGuidelines.md]: Agreed.  I'll also add an enforcement bullet that a warning should be issued for a pointer cast that could be implicit as a broader version of the CppCoreChecker's warning against using a reinterpret_cast<void*>

**Comment by gdr-at-ms** [comment]
Looks good to me.  Bjarne may have more comments.

**Comment by AndrewPardoe** [comment]
Thank you, Zachary!

**Pull request merged**